### PR TITLE
Add project node on top of group by with only outer references

### PR DESCRIPTION
**Pull request by zellerh**
Consider the following query:
```
create table foo(a int);
create table bar(b int);
select * from foo where foo.a in (select distinct foo.a from bar);
```
The above query will get preprocessed as shown below:
```
Algebrized query:
+--CLogicalSelect
   |--CLogicalGet "foo" ("foo"), Columns: ["a" (0), "ctid" (1), "xmin" (2), "cmin" (3), "xmax" (4), "cmax" (5), "tableoid" (6), "gp_segment_id" (7)] Key sets: {[1,7]}
   +--CScalarSubqueryAny(=)["a" (0)]
      |--CLogicalGbAgg( Global ) Grp Cols: ["a" (0)][Global], Minimal Grp Cols: [], Generates Duplicates :[ 0 ]
      |  |--CLogicalGet "bar" ("bar"), Columns: ["b" (8), "ctid" (9), "xmin" (10), "cmin" (11), "xmax" (12), "cmax" (13), "tableoid" (14), "gp_segment_id" (15)] Key sets: {[1,7]}
      |  +--CScalarProjectList
      +--CScalarIdent "a" (0)

Algebrized preprocessed query:
+--CLogicalSelect
   |--CLogicalGet "foo" ("foo"), Columns: ["a" (0), "ctid" (1), "xmin" (2), "cmin" (3), "xmax" (4), "cmax" (5), "tableoid" (6), "gp_segment_id" (7)] Key sets: {[1,7]}
   +--CScalarBoolOp (EboolopAnd)
      |--CScalarCmp (=)
      |  |--CScalarIdent "a" (0)
      |  +--CScalarIdent "a" (0)
      +--CScalarSubqueryExists
         +--CLogicalGbAgg( Global ) Grp Cols: ["a" (0)][Global], Minimal Grp Cols: [], Generates Duplicates :[ 0 ]
            |--CLogicalGet "bar" ("bar"), Columns: ["b" (8), "ctid" (9), "xmin" (10), "cmin" (11), "xmax" (12), "cmax" (13), "tableoid" (14), "gp_segment_id" (15)] Key sets: {[1,7]}
            +--CScalarProjectList
```
The CLogicalGbAgg has an outer reference in the grouping column, which
is not supported and causes us to crash in the retail build.

The fix is to add a project node below the GbAgg that echoes the outer
reference and to use that column as a grouping column:

```
Algebrized preprocessed query:
+--CLogicalSelect
   |--CLogicalGet "foo" ("foo"), Columns: ["a" (0), "ctid" (1), "xmin" (2), "cmin" (3), "xmax" (4), "cmax" (5), "tableoid" (6), "gp_segment_id" (7)] Key sets: {[1,7]}
   +--CScalarBoolOp (EboolopAnd)
      |--CScalarCmp (=)
      |  |--CScalarIdent "a" (0)
      |  +--CScalarIdent "a" (0)
      +--CScalarSubqueryExists
         +--CLogicalGbAgg( Global ) Grp Cols: ["ColRef_0016" (16)][Global], Minimal Grp Cols: [], Generates Duplicates :[ 0 ]
            |--CLogicalProject
            |  |--CLogicalGet "bar" ("bar"), Columns: ["b" (8), "ctid" (9), "xmin" (10), "cmin" (11), "xmax" (12), "cmax" (13), "tableoid" (14), "gp_segment_id" (15)] Key sets: {[1,7]}
            |  +--CScalarProjectList
            |     +--CScalarProjectElement "ColRef_0016" (16)
            |        +--CScalarIdent "a" (0)
            +--CScalarProjectList
```

Co-authored-by: Abhijit Subramanya <asubramanya@pivotal.io>
Co-authored-by: Hans Zeller <hzeller@pivotal.io>

### CODE REVIEW

File: libgpopt/src/operators/CExpressionPreprocessor.cpp [hunk 1]
```diff
@@ -579,14 +655,18 @@ CExpressionPreprocessor::PexprRemoveSuperfluousOuterRefs
...
+	const ULONG arity = newExpr->Arity();
 	CExpressionArray *pdrgpexprChildren = GPOS_NEW(mp) CExpressionArray(mp);
 	for (ULONG ul = 0; ul < arity; ul++)
 	{
-		CExpression *pexprChild = PexprRemoveSuperfluousOuterRefs(mp, (*pexpr)[ul]);
+		CExpression *pexprChild = PexprRemoveSuperfluousOuterRefs(mp, (*newExpr)[ul]);
 		pdrgpexprChildren->Append(pexprChild);
 	}
 
+	if (newExpr != pexpr)
```
Comment (ashuka24) [inline libgpopt/src/operators/CExpressionPreprocessor.cpp]: Could you please explain why this was added. I don't full understand the point of this check.

File: libgpopt/src/operators/CExpressionPreprocessor.cpp [hunk 2]
```diff
@@ -518,7 +525,9 @@ CExpressionPreprocessor::PexprRemoveSuperfluousOuterRefs
 			// the outer references are NOT the ONLY Group By column
 			//
 			// For example:
-			// -- Cannot remove t.b from groupby, because this will produce invalid plan
+			// -- Cannot remove t.b from groupby, because this will produce and invalid plan
```
Comment (chrishajas) [inline libgpopt/src/operators/CExpressionPreprocessor.cpp]: s/and/an

File: libgpopt/src/operators/CExpressionPreprocessor.cpp [hunk 3]
```diff
@@ -558,6 +569,71 @@ CExpressionPreprocessor::PexprRemoveSuperfluousOuterRefs
...
+
+				// build a new CLogicalGbAgg operator, with a new grouping columns list
+				CColRefArray *new_grouping_cols = GPOS_NEW(mp) CColRefArray(mp);
+				CExpression *new_projected_cols = (*projectExpr)[1];
+				for (ULONG ul = 0; ul < new_projected_cols->Arity(); ul++)
+				{
+					new_grouping_cols->Append(CUtils::PcrFromProjElem((*new_projected_cols)[ul]));
+				}
+				GPOS_ASSERT(NULL == popAgg->PdrgpcrArgDQA());
+				pop = GPOS_NEW(mp) CLogicalGbAgg(
```
Comment (chrishajas) [inline libgpopt/src/operators/CExpressionPreprocessor.cpp]: Do we need to release the previous pop?

File: libgpopt/src/operators/CExpressionPreprocessor.cpp [hunk 4]
```diff
@@ -558,6 +569,71 @@ CExpressionPreprocessor::PexprRemoveSuperfluousOuterRefs
...
+				CColRefArray *grouping_cols = popAgg->Pdrgpcr();
+				CExpression *child = (*pexpr)[0];
+				CExpressionArray *grouping_cols_arr = GPOS_NEW(mp) CExpressionArray(mp);
+
+				GPOS_ASSERT(0 < grouping_cols->Size());
+
+				// convert the outer reference(s) into a CExpressionArray
+				for (ULONG ul = 0; ul < grouping_cols->Size(); ++ul)
+				{
+					grouping_cols_arr->Append(CUtils::PexprScalarIdent(mp, (*grouping_cols)[ul]));
```
[the remark this answers is not in the document]
Comment (zellerh) [reply libgpopt/src/operators/CExpressionPreprocessor.cpp]: Thanks, great idea.

**Comment by zellerh** [comment]
Thanks for pointing out the similar code, @vraghavan78. Abhijit and I looked at that code but could unfortunately not find a good way to combine them, since there are also quite a few differences. Our code already uses method `CUtils::PexprAddProjection()`, which handles much of the code that would be common with the `CExpressionPreprocessor::PexprProjBelowSubquery()` method Venky referred to.

File: libgpopt/src/operators/CExpressionPreprocessor.cpp [hunk 5]
```diff
@@ -558,6 +569,65 @@ CExpressionPreprocessor::PexprRemoveSuperfluousOuterRefs
...
+				// add a project node on top of our child
+				CExpression *projectExpr = CUtils::PexprAddProjection(
+																	  mp,
+																	  child,
+																	  grouping_cols_arr,
+																	  false // don't add to hash table,
+																			// this is done at the end
+																			// of preprocessing
+																	 );
+				child->AddRef();
```
Comment (hardikar) [inline libgpopt/src/operators/CExpressionPreprocessor.cpp]: pattern is to generally addref before the call.

**Pull request merged**